modifiers/allocation/modifier.py

```diff
@@ -0,0 +1,204 @@
+# Ramble expects n_ranks and n_nodes to be set to *something* so even
+# if we want to fill those in ourselves, we have to supply something.
+SENTINEL_UNDEFINED_VALUE_INT = 7
+SENTINEL_UNDEFINED_VALUE_STR = "placeholder"
+
+
+def defined_allocation_options(expander):
+    """For each possible allocation option, check if it was filled in by
+    Ramble.
+    """
+    defined = {}
+    for alloc_opt in AllocOpt:
+        var_def = expander.expand_var(f"{{{alloc_opt.name.lower()}}}")
+        try:
+            val = AllocOpt.as_type(alloc_opt, var_def)
+        except ValueError:
+            continue
+
+        if val is not None:
+            defined[alloc_opt] = val
+
+    return defined
+
+
+class AttrDict(dict):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self["_attributes"] = set()
+
+    def __getattr__(self, *args, **kwargs):
+        return self.__getitem__(*args, **kwargs)
+
+    def __setattr__(self, *args, **kwargs):
+        self.__setitem__(*args, **kwargs)
+
+    def __delattr__(self, *args, **kwargs):
+        self.__delitem__(*args, **kwargs)
+
+    def __setitem__(self, key, value):
+        super().__setitem__(key, value)
+        if key != "_attributes":
+            self["_attributes"].add(key)
+
+    def defined(self):
+        return list((k, self[k]) for k in self["_attributes"])
+
+    @staticmethod
+    def from_predefined_variables(var_defs):
+        v = AttrDict()
+        for alloc_opt in AllocOpt:
+            setattr(v, alloc_opt.name.lower(), var_defs.get(alloc_opt, None))
+
+        return v
+
+
+class Allocation(BasicModifier):
+
+    name = "allocation"
+
+    tags("infrastructure")
+
+    # Currently there is only one mode. The only behavior supported right
+    # now is to attempt to request "enough" resources for a given
+    # request (e.g. to make sure we request enough nodes, assuming we
+    # know how many CPUs we want)"
+    mode("standard", description="Standard execution mode for allocation")
+    default_mode("standard")
+
+    def inherit_from_application(self, app):
+        super().inherit_from_application(app)
+
+        var_defs = defined_allocation_options(app.expander)
+        v = AttrDict.from_predefined_variables(var_defs)
+
+        # Calculate unset values (e.g. determine n_nodes if not set)
+        self.determine_allocation(v)
+
+        self.determine_scheduler_instructions(v)
+
+        # Definitions
+        for var, val in v.defined():
+            app.define_variable(var, str(val))
+
+    def determine_allocation(self, v):
+        if not v.n_ranks:
+            if v.n_ranks_per_node and v.n_nodes:
+                v.n_ranks = v.n_nodes * v.n_ranks_per_node
+
+        if not v.n_nodes:
+            if v.n_ranks:
+                cpus_request_per_task = 1
+                multi_cpus_per_task = v.n_cores_per_task or v.n_threads or 0
+                cpus_request_per_task = max(multi_cpus_per_task, 1)
```

Comment:
what is "task" here?

Reply:
I use "task" and "rank" interchangeably.

In either case I mean "a process that represents a running instance of a program" and I thought task was slightly more generic (could account for cases where we want to run processes w/o MPI).

Reply:
Unfortunately "task" is an overloaded term in HPC, where "rank" is unambiguous, so we should just use "rank" everywhere.

Reply:
That is now done